cli/flox-watchdog/src/main.rs

```diff
@@ -187,16 +177,38 @@ fn run(args: Cli) -> Result<(), Error> {
 // If the activation for a watchdog gets removed from the registry as stale by a different watchdog,
 // multiple watchdogs could perform cleanup.
 // The following can be run multiple times without issue.
-fn cleanup(socket_path: impl AsRef<Path>) {
+fn cleanup(
+    socket_path: impl AsRef<Path>,
+    activations_json_path: impl AsRef<Path>,
+    activation_id: impl AsRef<str>,
+) -> Result<()> {
     debug!("running cleanup");
-    let socket_path = socket_path.as_ref();
-    if socket_path.exists() {
-        if let Err(err) = process_compose_down(socket_path) {
-            error!(%err, "failed to run process-compose shutdown command");
+
+    let (activations_json, _lock) = read_activations_json(activations_json_path)?;
+    let Some(mut activations_json) = activations_json else {
+        bail!("watchdog shouldn't be running when activations.json doesn't exist");
+    };
+    activations_json.remove_activation(activation_id);
+
+    // Even if this activation has no more attached PIDs, there may be other
+    // activations for a different build of the same environment
+    if activations_json.is_empty() {
+        let socket_path = socket_path.as_ref();
+        if socket_path.exists() {
+            if let Err(err) = process_compose_down(socket_path) {
+                error!(%err, "failed to run process-compose shutdown command");
+            }
+        } else {
+            debug!(reason = "no socket", "did not shut down process-compose");
         }
-    } else {
-        debug!(reason = "no socket", "did not shut down process-compose");
     }
+
+    // We want to hold the lock until services are cleaned up
+    drop(_lock);
+
+    // TODO: remove activation state dir
```

Comment:
I this a "todo but 🤷🏼‍♂️" or do we need an issue for this?

Reply:
Went ahead and added it